# 01. Importing Libraries

In [26]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [27]:
# Import the data set into Jupyter as a new dataframe: df_ords_prior
path = r'/Users/ChuahLH/Desktop/Documents/Instacart Basket Analysis'

In [35]:
# Import the orders_products_merged dataframe
orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))

In [41]:
# Investigate the shape of the df_orders_products_merged
orders_products_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  order_dow  \
0             0   2539329        1             1          2   
1             1   2398795        1             2          3   
2             2    473747        1             3          3   
3             3   2254736        1             4          4   
4             4    431534        1             5          4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0        0.0           195         Soda        77              7     9.0   
1        1.0           195         Soda        77              7     9.0   
2        1.0           195         Soda        77              7     9.0   
3        1.0           195         Soda        77              7     9.0   
4        1.0           195         Soda        77              7     9.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [42]:
# Only work with a subset of the dataframe for the first one million rows
df = orders_products_merged[:1000000]

In [44]:
df.shape

(1000000, 16)

# 03. Define Price_Label & Busiest_Day

In [45]:
# define function to sort price ranges

def price_label(row):
    if row['prices'] <= 5: 
        return 'Low-range product'
    elif (row['prices']>5) and (row['prices']<=15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: 
        return 'Not enough data'

In [46]:
# Apply the function
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/37/dnh2n8_x6xx8nwj_pcczdql00000gn/T/ipykernel_7276/902492192.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [47]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [48]:
df['prices'].max()

14.8

In [49]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/37/dnh2n8_x6xx8nwj_pcczdql00000gn/T/ipykernel_7276/1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [50]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [51]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [52]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [53]:
orders_products_merged.loc[orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [54]:
orders_products_merged.loc[(orders_products_merged['prices'] <= 15) & (orders_products_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [55]:
orders_products_merged.loc[orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [56]:
orders_products_merged['price_range_loc'].value_counts(dropna = True)

Mid-range product     21861997
Low-range product     10126366
High-range product      417678
Name: price_range_loc, dtype: int64

In [57]:
# Renaming column from 'order_dow' to 'orders_day_of_the_week'. 
orders_products_merged = orders_products_merged.rename(columns={'order_dow': 'orders_day_of_week'})
orders_products_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0        0.0           195         Soda        77              7     9.0   
1        1.0           195         Soda        77              7     9.0   
2        1.0           195         Soda        77              7     9.0   
3        1.0           195         Soda        77              7     9.0   
4        1.0           195         Soda        77              7     9.0   

  _merge    price_range_loc  
0   both  Mid-range product  
1   both  Mid-range product  
2   both  Mid-range product  
3   both  Mid-range product  
4   both  Mid-range product

In [58]:
orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: orders_day_of_week, dtype: int64

In [59]:
# define function to sort busiest days
result = []

for value in orders_products_merged['orders_day_of_week']:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [60]:
orders_products_merged['busiest_day'] = result

In [61]:
orders_products_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22417684
Busiest day        6204404
Least busy         3783953
Name: busiest_day, dtype: int64

# 04. Task

In [66]:
## 2. Busiest days
# Checking the busiest days of the week
result_2 = []

for value in orders_products_merged['orders_day_of_week']:
  if value in [0,1]:
    result_2.append("Busiest days")
  elif value in [3,4]:
    result_2.append("Least busy days")
  else:
    result_2.append("Regularly busy days")

In [67]:
orders_products_merged['busiest_days'] = result_2
orders_products_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0        0.0           195         Soda        77              7     9.0   
1        1.0           195         Soda        77              7     9.0   
2        1.0           195         Soda        77              7     9.0   
3        1.0           195         Soda        77              7     9.0   
4        1.0           195         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day         busiest_days  
0   both  Mid-range product  Regularly busy  Regularly busy days  
1   both  Mid-range product      Least busy      Least busy days  
2   both  Mid-range product      Least busy      Least busy days  
3   both  Mid-range product      Least busy      Least busy days  
4   both  Mid-range product      Least busy      Least busy days

In [68]:
orders_products_merged['busiest_days'].value_counts()

Regularly busy days    12916527
Busiest days           11864860
Least busy days         7624654
Name: busiest_days, dtype: int64

In [69]:
# Checking result against output of business per day by the sum of the according single days
orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204404
1    5660456
6    4496635
2    4213986
5    4205906
3    3840701
4    3783953
Name: orders_day_of_week, dtype: int64

In [70]:
# Cross check busiest days
busiest_days = 6204404 + 5660456

In [71]:
print(busiest_days)

11864860


In [72]:
# Cross check least busiest days
least_busiest_days = 3840701 + 3783953

In [73]:
print(least_busiest_days)

7624654


Cross check against the output and both matches. The busiest days is 11864860 and the least busiest days is 7624654.

In [74]:
## Find the Busiest hours
# Displaying orders per hour descending by orders
orders_products_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761862
11    2736242
14    2689251
15    2662246
13    2661057
12    2618613
16    2535313
9     2454273
17    2087721
8     1718182
18    1636566
19    1258352
20     976182
7      891082
21     795658
22     634247
23     402335
6      290500
0      218786
1      115706
5       87964
2       69377
4       53245
3       51281
Name: order_hour_of_day, dtype: int64

In [75]:
# Creating the function for busiest hours of the day by periods
result_3 = []

for value in orders_products_merged['order_hour_of_day']:
  if value>=9 and value<=16:
    result_3.append("Most orders")
  elif value<=6:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [76]:
# Creating new column 'busiest_period_of_day' with the results of the for loop
orders_products_merged['busiest_period_of_day'] = result_3
orders_products_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   

   reordered  Unnamed: 0_y product_name  aisle_id  department_id  prices  \
0        0.0           195         Soda        77              7     9.0   
1        1.0           195         Soda        77              7     9.0   
2        1.0           195         Soda        77              7     9.0   
3        1.0           195         Soda        77              7     9.0   
4        1.0           195         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product      Least busy      Least busy days   
2   both  Mid-range product      Least busy      Least busy days   
3   both  Mid-range product      Least busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

In [77]:
# Counting values of new column by determined periods
orders_products_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118857
Average orders    10400325
Fewest orders       886859
Name: busiest_period_of_day, dtype: int64

In [78]:
# Exporting data frame as pickle file to prepared data folder
orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))